File: Crystal/Math/Cone3d.cpp
#include "Cone3d.h"

using namespace Crystal::Math;

Vector3dd Cone3d::getPosition(const double u, const double v, const double w) const
{
	const auto uu = u * 2.0 * PI;
	const auto r = w * radius * (1.0 - v);
	const auto x = r * ::cos(uu);
	const auto y = r * ::sin(uu);
	const auto z = v * height;
	return bottom + Vector3dd(x, y, z);
}

bool Cone3d::isInside(const Vector3dd& position) const
{
	assert(false);
	return false;
}

File: CGStudio/PGCore/Math/Primitive3d/Cone3d.cs
using System;

namespace PG.Core.Math
{
    public class Cone3d : ISurface3d, IVolume3d
    {
        public Cone3d()
        {
            Radius = 1.0;
            Height = 1.0;
            Bottom = new Vector3d(0, 0, 0);
        }

        public Cone3d(double radius, double height, Vector3d bottom)
        {
            Radius = radius;
            Height = height;
            Bottom = bottom;
        }

        public double Radius { get; }

        public double Height { get; }

        public Vector3d Bottom { get; }

        public Vector3d GetPosition(double u, double v)
        {
            return GetPosition(u, v, 1.0);
        }

        public Vector3d GetPosition(double u, double v, double w)
        {
            var uu = u * 2.0 * System.Math.PI;
            var r = w * Radius * (1.0 - v);
            var x = r * System.Math.Cos(uu);
            var y = r * System.Math.Sin(uu);
            var z = v * Height;
            return Bottom + new Vector3d(x, y, z);
        }

        public bool IsInside(Vector3d position)
        {
            throw new NotImplementedException();
        }
    }
}
